examples/genserver.exs

```diff
@@ -0,0 +1,125 @@
+Mix.install([:mint_web_socket, :castore])
+
+defmodule Ws do
+  use GenServer
+
+  require Logger
+  require Mint.HTTP
+
+  def start_link(_) do
+    GenServer.start_link(__MODULE__, [])
+  end
+
+  def send_message(pid, text) do
+    GenServer.call(pid, {:send_text, text})
+  end
+
+  @impl GenServer
+  def handle_call({:send_text, text}, _from, state) do
+    new_state = send_frame({:text, text}, state)
+    {:reply, :ok, new_state}
+  end
+
+  @impl GenServer
+  def init(_opts) do
+    {:ok, conn} = Mint.HTTP.connect(:http, "127.0.0.1", 1234)
+
+    {:ok, conn, ref} =
+      Mint.WebSocket.upgrade(
+        :ws,
+        conn,
+        "/",
+        []
+      )
+
+    # Receive response to the upgrade call above
+    receive do
+      message when Mint.HTTP.is_connection_message(conn, message) ->
+        {:ok, conn, responses} = Mint.WebSocket.stream(conn, message)
+        response = reduce_response(responses, ref)
+        {:ok, conn, websocket} = Mint.WebSocket.new(conn, ref, response.status, response.headers)
+
+        {:ok,
+         %{
+           ref: ref,
+           websocket: websocket,
+           conn: conn
+         }}
+    end
```

Comment:
I think I still prefer the `receive` block as it means there is no extra state needed for upgrading, and I can't think of a way in which it goes wrong, but I've replaced it with your suggestion.

Reply:
It's pretty contrived and I doubt you'd see it in the wild but if the server doesn't send the HTTP upgrade response, this `receive/1` will hang, and because it's in a `c:GenServer.init/1`, the `start_link/1` call will hang indefinitely. That could be addressed with an `after` timeout I suppose, but I prefer handling the timeout in the `GenServer.call/3`.

I think you're right that `receive/1` should work fine here though, and it does simplify a good amount the state-machine-like behavior of the GenServer 👍